### DISCUSSION TITLE
[Feature] Modulated tracker file playback

### DISCUSSION THREAD

**Discussion by Amidweiz** [ikemen-engine/Ikemen-GO#843] [Feature Request]
It would be nice if Ikemen GO plus would be able to play modulated tracker files natively rather than needing to convert them to mp3, ogg, etc. Such files as .MOD, .S3M, .XM, etc. Not sure if there is a version of OpenMPT for go lang or if openAL is able to play them natively, but it would be nice anyways. Looping is done inside the file so you wouldn't need to worry about loops with loopstart and loopend in the def file.

**Comment by GreatFunk** [comment]
i like that idea.

**Comment by SuperFromND** [comment]
There does appear to be a library for this specific purpose, [`gotracker/playback`](https://github.com/gotracker/playback). Doesn't appear to be especially active (last commit July 2022), and I'm not sure how easy it would be to integrate into the existing audio systems.

**Reply by SuperFromND** [reply]
Okay so I did briefly attempt to add this library into [our Beep fork](https://github.com/ikemen-engine/beep), but it seems like doing so would require adding a rather large number of dependencies AND it would've bumped version requirements to 1.21. Even ignoring that, I wasn't really able to figure out its (rather barren) documentation to get anything playing. Just seemed too different in terms of boilerplate flow compared to how we currently handle other audio formats, but maybe I'm just not very experienced yet.

In case anyone else wants to give it a try, here's some notes I took:
- To add support for new formats to Beep, create a new folder named after the format (I used `modtracker`), create a `decode.go`, and then from there you need to create a `Decode` function, a `Stream` function, and some other utility functions like `Seek` and `Len`. I'd recommend looking at a combination of `midi` and `mp3` to get an idea of what the code should generally look like.
- [`gotracker/playback`](https://github.com/gotracker/playback) is the actual library that handles playing tracker music. The examples they used seem to roll in their own bespoke library for playing back audio through standard out, [`gotracker/gomixing`](https://github.com/gotracker/gomixing), so you'll have to either just roll with that or find a way to sidestep it since Ikemen already has audio mixing/output of its own.
- To get Ikemen to recognize the file format, just edit `sound.go` around line 200 or so. In my attempt, it looked like this:
```go
// ...
	} else if HasExtension(bgm.filename, ".mid") || HasExtension(bgm.filename, ".midi") {
		if soundfont, sferr := loadSoundFont(audioSoundFont); sferr != nil {
			err = sferr
		} else {
			bgm.streamer, format, err = midi.Decode(f, soundfont)
			bgm.format = "midi"
		}
	} else if HasExtension(bgm.filename, ".mod") { // only focused on .mod, but .xm and .s3m should be trivial to add here
		bgm.streamer, format, err = modtracker.Decode(f)
		bgm.format = "modtracker"
	} else {
		err = Error(fmt.Sprintf("unsupported file extension: %v", bgm.filename))
	}
// ...
```
- Note that `format` string in the `bgm` struct. I'm not actually sure what it's used for, if anything at all. Maybe worth investigating or labelling that somewhere down the line.
